test/unittests/state_transition.cpp

```diff
@@ -0,0 +1,49 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2023 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "state_transition.hpp"
+
+namespace evmone::test
+{
+void state_transition::TearDown()
+{
+    auto& state = pre;
+    const auto res = evmone::state::transition(state, block, tx, EVMC_CANCUN, vm);
+    ASSERT_TRUE(holds_alternative<TransactionReceipt>(res))
+        << std::get<std::error_code>(res).message();
+    const auto& receipt = std::get<TransactionReceipt>(res);
+
+    EXPECT_EQ(receipt.status, expect.status);
+    if (expect.gas_used.has_value())
+    {
+        EXPECT_EQ(receipt.gas_used, *expect.gas_used);
+    }
+
+    for (const auto& [addr, expected_acc] : expect.post)
+    {
+        const auto acc = state.find(addr);
+        if (!expected_acc.exists)
+        {
+            // TODO: Add printing for address and bytes32.
+            EXPECT_EQ(acc, nullptr) << "account " << hex(addr) << " should not exist";
+        }
+        else
+        {
+            ASSERT_NE(acc, nullptr) << "account " << hex(addr) << " should exist";
+            if (expected_acc.nonce.has_value())
+            {
+                EXPECT_EQ(acc->nonce, *expected_acc.nonce);
+            }
+            if (expected_acc.balance.has_value())
+            {
+                EXPECT_EQ(acc->balance, *expected_acc.balance);
+            }
+            for (const auto& [key, value] : expected_acc.storage)
+            {
+                EXPECT_EQ(acc->storage.at(key).current, value);
+            }
```

Comment:
Should this also check for extra values in account storage that are not listed in expected?

Reply:
Not sure. For accounts I wanted to use `std::nullopt` to mark an account which should not exist. For storage we have null value. I guess we can extend the checks as you suggested until proven annoying.

Reply:
Added this for now.

Same could be done for accounts and possible eliminating `.exists` flag.